File: Turnstile.Scala/src/main/scala/actors/TurnstileActor.scala
package actors

import akka.actor.LoggingFSM
import messages.{InsertCoin, PushBar}

/**
  * Created by gianlo.fagiolo on 10/06/2016.
  */

class TurnstileActor extends LoggingFSM[TurnstileStatus, Coins] {
  startWith(Locked, Coins.noCoin)
  when(Locked) {
    case Event(InsertCoin, _) => goto(Unlocked).using(stateData.addOne)
    case Event(PushBar, _) => {
      println(s"Seriously Buddy? Before you can pass you need to insert a coin! I'm currently $stateName.")
      stay
    }
  }

  when(Unlocked) {
    case _ => throw new NotImplementedError()
  }
}

sealed trait TurnstileStatus

case object Locked extends TurnstileStatus

case object Unlocked extends TurnstileStatus

case class Coins(total: Int) {
  def addOne = Coins(total + 1)
}

object Coins {
  def noCoin = Coins(0)
}

File: Turnstile.CSharp/Turnstile.CSharp/Actors/TurnstileActor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Akka.Actor;

namespace Turnstile.CSharp.Actors
{
    public enum TurnstileStatus
    {
        Locked,
        Unlocked
    }

    public class Coins
    {
        public int Total { get; }

        private Coins(int total)
        {
            Total = total;
        }

        public Coins AddOne()
        {
            return new Coins(Total + 1);
        }

        public static Coins NoCoins()
        {
            return new Coins(0);
        }
    }

    class TurnstileActor: FSM<TurnstileStatus, Coins>
    {
        public TurnstileActor()
        {
            InitializeFsm();
        }

        private void InitializeFsm()
        {
            StartWith(TurnstileStatus.Locked, Coins.NoCoins());
            When(TurnstileStatus.Locked, LockedLogic);
            When(TurnstileStatus.Unlocked, UnlockedLogic);
        }

        private State<TurnstileStatus, Coins> UnlockedLogic(Event<Coins> fsmevent)
        {
            throw new NotImplementedException();
        }

        private State<TurnstileStatus, Coins> LockedLogic(Event<Coins> fsmevent)
        {
            throw new NotImplementedException();
        }
    }
}
